website/notifications/events/utils.py

```diff
@@ -0,0 +1,138 @@
+    :param user: User instance who started the event
+    :param source_event: <guid>_event_name
+    :param source_node: node from where the event happened
+    :param event: new guid event name
+    :param node: node where event ends up
+    :return: Moved, to be warned, and removed users.
+    """
+    remove = utils.users_to_remove(source_event, source_node, node)
+    source_node_subs = compile_subscriptions(source_node, utils.find_subscription_type(source_event))
+    new_subs = compile_subscriptions(node, utils.find_subscription_type(source_event), event)
+
+    # Moves users into the warn bucket or the move bucket
+    move = subscriptions_users_union(source_node_subs, new_subs)
+    warn = subscriptions_users_difference(source_node_subs, new_subs)
+
+    # Removes users without permissions
+    warn, remove = subscriptions_node_permissions(node, warn, remove)
+
+    # Remove duplicates
+    warn = subscriptions_users_remove_duplicates(warn, new_subs, remove_same=False)
+    move = subscriptions_users_remove_duplicates(move, new_subs, remove_same=False)
+
+    # Remove duplicates between move and warn; and move and remove
+    move = subscriptions_users_remove_duplicates(move, warn, remove_same=True)
+    move = subscriptions_users_remove_duplicates(move, remove, remove_same=True)
+
+    for notifications in constants.NOTIFICATION_TYPES:
+        # Remove the user who started this whole thing.
+        user_id = user._id
+        if user_id in warn[notifications]:
+            warn[notifications].remove(user_id)
+        if user_id in move[notifications]:
+            move[notifications].remove(user_id)
+        if user_id in remove[notifications]:
+            remove[notifications].remove(user_id)
+
+    return move, warn, remove
+
+
+def subscriptions_node_permissions(node, warn_subscription, remove_subscription):
+    for notification in constants.NOTIFICATION_TYPES:
+        subbed, removed = utils.separate_users(node, warn_subscription[notification])
+        warn_subscription[notification] = subbed
+        remove_subscription[notification].extend(removed)
+        remove_subscription[notification] = list(set(remove_subscription[notification]))
+    return warn_subscription, remove_subscription
+
+
+def subscriptions_users_union(emails_1, emails_2):
+    return {
+        notification:
+            list(
+                set(emails_1[notification]).union(set(emails_2[notification]))
+            )
+        for notification in constants.NOTIFICATION_TYPES.keys()
+    }
+
+
+def subscriptions_users_difference(emails_1, emails_2):
+    return {
+        notification:
+            list(
+                set(emails_1[notification]).difference(set(emails_2[notification]))
+            )
+        for notification in constants.NOTIFICATION_TYPES.keys()
+    }
+
+
+def subscriptions_users_remove_duplicates(emails_1, emails_2, remove_same=False):
+    emails_list = dict(emails_1)
+    for notifications in constants.NOTIFICATION_TYPES:
```

Comment:
You may be able to refactor this using itertools.combinations: https://docs.python.org/2/library/itertools.html#itertools.combinations